### PR TITLE
[FX] Experimental type annotation pass using Python signatures

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53771 [FX] Normalize torch. namespace ops
* **#53653 [FX] Experimental type annotation pass using Python signatures**
* #53770 [WIP] Function to retrieve inspect.Signature instances for PyTorch ops
 
    Use Python function signatures to annotate types for `Nodes` within an FX graph.
    This pulls out Python function signatures for both standard `torch.nn` Module

Differential Revision: [D26924485](https://our.internmc.facebook.com/intern/diff/D26924485)

### CODE REVIEW

File: torch/fx/experimental/schema_type_annotation.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    def call_function(self, target : Target, args : Tuple[Argument, ...], kwargs : Dict[str, Any]):
+        python_ret_type = None
+        if self.annotate_functionals and target.__module__ == 'torch.nn.functional':
+            target_for_analysis = target
+            if target in boolean_dispatched:
+                # HACK: `boolean_dispatch` as used in `torch.nn.functional` makes it so that we have
+                # a 2-way dispatch based on a boolean value. Here we check that the `true` and `false`
+                # branches of the dispatch have exactly the same signature. If they do, use the `true`
+                # branch signature for analysis. Otherwise, leave this un-normalized
```
Comment (Chillee) [inline torch/fx/experimental/schema_type_annotation.py:43]: Shouldn't be "un-normalized"?

File: torch/fx/experimental/schema_type_annotation.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            args (Tuple): Arguments that appear at the callsite for `target`
+            kwargs (Dict): Keyword arugments that appear at the callsite for `target`
+
+        Returns:
+
+            Optional[Dict]: Normalized kwargs for `target`, or `None` if this target is not
+                supported
+        """
+        assert not isinstance(target, str)
+        sig = inspect.signature(inspect.unwrap(target))
```
Comment (Chillee) [inline torch/fx/experimental/schema_type_annotation.py:105]: Is this guaranteed to work/not throw an error?

File: torch/fx/experimental/schema_type_annotation.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        if self.annotate_get_attrs:
+            module_itr = self.module
+            assert isinstance(target, str)
+            atoms = target.split('.')
+            for i, atom in enumerate(atoms):
+                if not hasattr(module_itr, atom):
+                    raise RuntimeError(f'Node referenced nonextent target {".".join(atoms[:i])}!')
+                module_itr = getattr(module_itr, atom)
+
+            attr_proxy.node.type = type(module_itr)
```
Comment (Chillee) [inline torch/fx/experimental/schema_type_annotation.py:83]: Is this type the same "form" as what `inspect` provides, and is that a problem? For example, if you have a list of integers what happens in `getattr` vs `call_module`?

**Review by Chillee — APPROVED** [review]
Mostly LGTM

File: torch/fx/experimental/schema_type_annotation.py (hunk 1 above)
Comment (jamesr66a) [reply torch/fx/experimental/schema_type_annotation.py:43]: Don't really understand the comment?

File: torch/fx/experimental/schema_type_annotation.py (hunk 3 above)
Comment (jamesr66a) [reply torch/fx/experimental/schema_type_annotation.py:83]: Good point. Seems pretty annoying to solve. I might rebase so that I'm on top of https://github.com/pytorch/pytorch/pull/53770/files#diff-e7602003c853555c44d805885670cf3500aefe9b8cfec7f8571e2d11f57fd56dR31 and I can use `torch._C._jit_try_infer_type` and feed that result into `_torchscript_type_to_python_type`

File: torch/fx/experimental/schema_type_annotation.py (hunk 1 above)
Comment (Chillee) [reply torch/fx/experimental/schema_type_annotation.py:43]: Like, this comment is copied from `normalize.py`, where the goal is to normalize the args/kwargs. Your comment here says to "otherwise, leave this un-normalized". I don't think that makes sense in the context of this pass.
